adding Nathan's SSO/Custom IDP docs (#625)
chainguard-dev/edu: content/chainguard/chainguard-enforce/authentication/custom-idps/index.md

---

> title : "Using Custom Identity Providers to Authenticate to Chainguard"

Author: I agree, I wasn't sure how to bucket this. I think I'll put the assumed identities doc in this same bucket, so that will hopefully make it feel more fleshed out

* The `authorization code` grant type (sometimes called the `authorization code` *flow*).
* The standard `openid`, `email` and `profile` scopes. Note that the Chainguard platform will partially function with only the `openid` scope, but full functionality requires the `email` and `profile` scopes as well.

Customer-managed identity providers must also have a public, unauthenticated OIDC discovery endpoint.

In order to set up SSO for your identity provider, you must configure an OIDC application that the Chainguard platform can use on your identity provider. Following that, you have to configure the Chainguard platform to use that application.

### Integration Guides for supported identity providers

If your identity provider is Okta, Ping Identity or Azure Active Directory, we’ve published step-by-step integration guides for your platform.

* [Okta](/chainguard/chainguard-enforce/authentication/example-idps/okta/)
* [Ping Identity](/chainguard/chainguard-enforce/authentication/example-idps/ping-id/)
* [Azure Active Directory](/chainguard/chainguard-enforce/authentication/example-idps/azure-ad/)

> If you aren't using one of these identity providers, you can complete the following Generic Integration Guide to configure your provider to work with Chainguard Enforce.

Author: LMK what you think of what I did here. I tried it as a separate note but it felt a little wonky, so I made the warning part of the paragraph following the list

> If you aren't using one of these identity providers, you can complete the following Generic Integration Guide to configure your provider to work with Chainguard Enforce. However, be aware that Chainguard does not actively support identity providers other than the ones listed previously. If you'd like to request that we support a specific identity provider, we encourage you to [contact us](https://www.chainguard.dev/contact?utm_source=docs) to do so.

Reviewer: I think this is good, I would maybe even make fewer claims, like below, but feel free to finesse further!

```suggestion
If you aren't using one of these identity providers, you can complete the following Generic Integration Guide to configure your provider to work with Chainguard Enforce. However, be aware that Chainguard does not actively support identity providers other than the ones listed previously. If you are using an alternate identity provider, we encourage you to [contact us](https://www.chainguard.dev/contact?utm_source=docs) to learn more.
```

### Generic Integration Guide

For a generic OIDC-compatible identity provider, start by creating an OIDC application. If possible, set as much metadata as possible for the application so that your users can identify this application as the Chainguard platform. The following assets and details can be helpful to include in metadata.

* The Console homepage is [https://console.enforce.dev/](https://console.enforce.dev)
* Our terms of service can be found at [chainguard.dev/terms-of-service](https://www.chainguard.dev/terms-of-service)
* Our terms of use can be found at [chainguard.dev/terms-of-use](https://www.chainguard.dev/terms-of-use)
* Our privacy policy is located at [chainguard.dev/privacy-notice](https://www.chainguard.dev/privacy-notice)
* You can also add a Chainguard logo icon here to help your users visually identify this integration. The icon from the [Chainguard Enforce Console](https://console.enforce.dev/logo512.png) will be suitable for most platforms.

Next, configure your OIDC application as follows:

* Set redirect URI to `https://issuer.enforce.dev/oauth/callback`

> * Restrict grants types to **authorization code** only. It is critical that your application does not support "client credentials", "device code", "implicit" or other grant types (sometimes called “flows”).

Author: Yes, yes it should

The IAM capabilities `identity_providers.create`, `identity_providers.update`, `identity_providers.list` and `identity_providers.delete` control which users can read and manipulate identity providers. The built-in roles `viewer`, `editor` and `owner` have the following identity provider related capabilities.

| **Role** | **Capabilities** |
|----------|----------|
| `viewer`   | `identity_providers.list`   |
| `editor`   | `identity_providers.list`   |
| `owner`   | `identity_providers.create`, `identity_providers.list`, `identity_providers.update`, `identity_providers.delete`   |

As with other capabilities, if an identity provider capability is granted at a parent IAM group, it is inherited on all child IAM groups below the parent.

> ## Break-glass accounts

Author: That's a very good point. I'm going to give "backup account" a try.
